docs/config.md

```diff
@@ -0,0 +1,51 @@
+# Configuration
+The philosophy we will use when exposing HCO configuration options is that only
+options that all operators consume can be exposed on the default HCO CR.
+Configuration options that are for specific component operators, will be hidden
+but accepted as valid configuration.
+
+Exposed variables are expressed in the HCO [CSV file](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/dfdd4ac492c1d91c130eec03af1e4f6b04d54c7e/deploy/converged/olm-catalog/kubevirt-hyperconverged/0.0.1/kubevirt-hyperconverged-operator.v0.0.1.clusterserviceversion.yaml#L18-L20)
+```yaml
+          "spec": {
+            "CDIImagePullPolicy":"IfNotPresent",
+            "KubeVirtImagePullPolicy":"IfNotPresent"
+          }
+```
+
+## 1.0 Config Options
+All component operators must express each of the HCO `Config Options` on their CR.
+The component variable:
+  - does __not__ have to be the same name as the HCO variable
+  - can be expressed by multiple granular variables
```

Comment:
meaning that we can have different registry? quay vs docker. the same for versioning

Reply:
I was thinking there could be multiple images that your operator manages.  Each of those images can have their own variable name, but the registry for those images will come from the HCO.

Reply:
That won't work for us on U/S. Some images are not owned by us but by their respective projects.

Reply:
I think in that case, we'll have the HCO set the requested registry on the CR, but when pulling the image fails (which it always will), your operator can fall back to the upstream registry.

Reply:
IIUIC that would require us to implement extra logic, monitor deployed pods for specific failure and then redeploy them with a different registry. That sounds little too complicated.

What would be the use for this feature? Shouldn't each operator be responsible for it? If HCO sets other operator as dependencies, those other operators should already be set up with the right registry for the cluster they run it.

Reply:
The HCO is responsible for configuring components.  It's going to tell component operators what application they are deploying and managing.  What I'm proposing is that in this case, the operator shouldn't listen to the HCO's config because it knows better. 

> What would be the use for this feature? 
>
The registry your operator pulls images from should be configurable to allow someone to pull any downstream or internal registry.

> Shouldn't each operator be responsible for it? If HCO sets other operator as dependencies, those other operators should already be set up with the right registry for the cluster they run it.
>
Component operators aren't responsible for providing the registry when glued together with the HCO.  We want to have one source of truth when we glue together all the components.

Reply:
Thanks, I think I understand now.

We ship 5 images now, each from a different registry. I see 3 choices here:

- Configure/override registry for every component from HCO
- Configure/override registry for all network components
- Configure/override registry for all components (this one is proposed in the document, right?)

I'm fine with all of them, but I don't want to introduce special handling on network operator side (try to deploy an image and fall back to default if it fails to pull can bring unexpected results and would also require some rollback/upgrade support on our side). Could we make it so if a user explicitly sets the registry, they are responsible of shipping all components there?

Reply:
> Configure/override registry for all components (this one is proposed in the document, right?)
> 
Yes, this is what's proposed.

> Could we make it so if a user explicitly sets the registry, they are responsible of shipping all components there?
>
If that's works for you guys from an upstream perspective, then let's go with that assumption.

Reply:
A variable for preferred "product" registry would probably work for most of our container management CRs. The operator can they obey it or not depending on what is really needed.

The rest of the variables (image name, pull policy) should stay under the control of the operator I think.